draft-ietf-jsonpath-base.md

```diff
@@ -1104,6 +1094,23 @@ Some examples:
 | `[3,4,5]` | `$[?index(@)==2]`<br>`$[?index(@)==17]` | `[5]`<br>`[]` | Select array element |
 | `{"a":{"b":{5},c:0}}` | `$[?@.b==5 && !@.c]` | `[{"b":{5},c:0}]` | Existence  |
 
+
+## Normalized Paths
+
+A Normalized Path identifies a node by providing a query that results in exactly that node. For example,
+the JSONPath expression `$.book[?(@.price<10)]` could select two values with Normalized Paths
+`$.book[3]` and `$.book[5]`.
+
+An example application of Normalized Paths is a JSONPath implementation which outputs Normalized Paths
+instead of, or in addition to, the values identified by these paths.
+
+Since bracket notation is more general than dot notation, it is convenient in constructing
+Normalized Paths. However, Normalized Paths are not required to use bracket notation when dot notation is appropriate.
+
+Normalized Paths are not necessarily canonical since two distinct Normalized Paths may identify the same node.
+For example, `$.a` and `$['a']` are equivalent Normalized Paths. With certain JSON documents, `$[-1]` and `$[4]` may
+be equivalent Normalized Paths.
+
```

Comment:
Here
```
$['a']
$[4]
```
qualifies for a normalized Path only.

Reply:
Please see ...
```
normalized-path = root-selector *("[" (quoted-member-name / positiv-element-index) "]")

positive-element-index = ( "0" / (DIGIT1 *DIGIT) )
```

from above ...

Reply:
s/positiv/unsigned/  (0 is allowed, but not positive)

Reply:
... I wanted to avoid `non-negative-element-index` ... now I like `unsigned` ...

Reply:
> ... I wanted to avoid `non-negative-element-index` ... now I like `unsigned-element-index` ...

I think `non-negative-element-index` is better. The domain of indices is signed integers. That domain is restricted to non-negative integers for normalized paths, it's not that they've suddenly become unsigned.

Reply:
> Please see ...
> 
> ```
> normalized-path = root-selector *("[" (quoted-member-name / positiv-element-index) "]")
> 
> positive-element-index = ( "0" / (DIGIT1 *DIGIT) )
> ```
> 
> from above ...

Unless the mailing list pushes back, I plan to adopt this prescriptive approach.

Reply:
> > ... I wanted to avoid `non-negative-element-index` ... now I like `unsigned-element-index` ...
> 
> I think `non-negative-element-index` is better. The domain of indices is signed integers. That domain is restricted to non-negative integers for normalized paths, it's not that they've suddenly become unsigned.

I get this discussion often.  Non-negative is pedantic.  The range of actual array indexes actually is unsigned integer, so why not just say that.  (We also allow indexing with negative integers, but that means something different.)

Reply:
> > I think `non-negative-element-index` is better. The domain of indices is signed integers. That domain is restricted to non-negative integers for normalized paths, it's not that they've suddenly become unsigned.
> 
> I get this discussion often. Non-negative is pedantic. The range of actual array indexes actually is unsigned integer, so why not just say that. (We also allow indexing with negative integers, but that means something different.)

In this context, "pedantic" is just another name for "clear". It's unclear what it means to mix "unsigned" and "negative" integers on the same domain. It doesn't translate well into programming languages.

Reply:
> Please see ...
> 
> ```
> normalized-path = root-selector *("[" (quoted-member-name / positiv-element-index) "]")
> 
> positive-element-index = ( "0" / (DIGIT1 *DIGIT) )
> ```
> 
> from above ...

`quoted-member-name` allows two sort of variability: single or double quoted strings and escapes or unescaped characters. Should we leave these open or attempt to tie them down so there is one and only one representation of any Normalized Path?

Reply:
I sensed the direction of the discussion that we wanted to have deterministic Normalized Path results, i.e. two implementations would generate the same output paths for the same JSON value and query.  That would need us to pick the kind of quoting and the details of the escaping.  Deterministic results generally help in testing.  We might still decide that this is too much work.

Reply:
Please see my correction above ...

> hmm ... quoted-member-name must be replaced by single-quoted-member-name above.

so it should read somewhat like ...
```
normalized-path = root-selector *("[" *(%x27 *single-quoted %x27) / positiv-element-index) "]")

unsigned-element-index = ( "0" / (DIGIT1 *DIGIT) )
```

Reply:
@cabo: I think, the benefit of clean testing justifies the effort here. However, I cannot really judge the effort involved in documenting escaping.

Reply:
Ok, I suggest we mandate single quotes (to align with JSON) and escapes (except for \/ and \uxxxx, instead of which / and the unicode character for U+xxxx must be used, respectively). Is that a reasonable starting position?

Reply:
... but, isn't escaping already clearly documented with `single-quoted` already in the current draft ... ?

Reply:
> ... but, isn't escaping already clearly documented with `single-quoted` already in the current draft ... ?

No, because it permits at least two representations of any given character. I think we should try to nail this down so there is a unique representation.

Reply:
To be clear, JSON uses double quotes, so we are using single quotes to evade further JSON escaping, except where we need to escape.  We can't use the actual characters for \u0000 to \u001f, so we should use \b\t\n\f\r or the \u00xx for the other 11 cases.  We also need `\\` and `\'`, but we don't need `\/`.  I'm sure there will be people who would want to escape anything including and above \u00ff as well.  Oh, and we need to specify the case of the hexchars.

Reply:
Gotcha.

> I'm sure there will be people who would want to escape anything including and above \u00ff as well.

I guess escaping this range is safer, even if it can be criticised as ASCII-centric. It wouldn't surprise me if mandating unescaped characters for that range caused problems in some older programming languages.

Reply:
I was thinking a while about that escape thing. There is a strong relation to string comparison, see ...

https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/117#issuecomment-909356265

Consensus here obviously was *code point* i.e. *byte string* comparison.

https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/117#issuecomment-1014932216

In this case here we need not to deal with arbitrary text content, but rather with member names. This should be less crtitical then. 

So, if restricting to code point equality here again, is all in all consistent.

I tried to collect some common scenarios:

1. Query results are reused on the same JSON source document for revisiting nodes with the purpose of ...
2. Normalized pathes are applied to similar JSON documents (output from a single application or into a standardised JSON format) by different users with different implementations on different platforms.
3. Compare query results from different implementations for test purposes.

Discussion:
1. Should be mostly applied in practise (estimated). Is uncritical regarding escapes, since JSON document, implementation, platform remains the same.
2. Is the most critical from my point of view. But the probability of dealing with escaped characters in this scenario is considered rather low.
3. For developers or advanced users only. Should be uncritical regarding escapes.

Are there other relevant scenarios, that I might be missing here?

So Glyn's approach or code point equality seems to be a pragmatic solution here.